test/device_selection_tests.cc

```diff
@@ -0,0 +1,641 @@
+#include "catch2/catch_test_macros.hpp"
+#include "catch2/generators/catch_generators.hpp"
+#include "catch2/matchers/catch_matchers_string.hpp"
+#include "spdlog/sinks/ostream_sink.h"
+#include "test_utils.h"
+#include <celerity.h>
+
+struct mock_platform;
+struct mock_device {
+	mock_device() : platform(nullptr), id(0), type(cl::sycl::info::device_type::gpu) {}
+
+	mock_device(size_t id, mock_platform& platform, cl::sycl::info::device_type type = cl::sycl::info::device_type::gpu)
+	    : platform(&platform), id(id), type(type) {}
+
+	mock_platform& get_platform() const { return *platform; }
+
+	template <cl::sycl::info::device Param>
+	std::string get_info() const {
+		return "bar";
+	}
+
+	bool operator==(const mock_device& other) const { return other.id == id; }
+
+	cl::sycl::info::device_type get_type() const { return type; }
+
+	size_t get_id() { return id; }
+
+  private:
+	mock_platform* platform;
+	size_t id;
+	cl::sycl::info::device_type type;
+};
+struct mock_platform {
+	mock_platform(size_t id) : id(id) {}
+
+	void set_devices(std::vector<mock_device> devices) { this->devices = devices; }
+
+	std::vector<mock_device> get_devices(cl::sycl::info::device_type type = cl::sycl::info::device_type::all) const {
+		if(type != cl::sycl::info::device_type::all) {
+			std::vector<mock_device> devices_with_type;
+			for(auto device : devices) {
+				if(device.get_type() == type) { devices_with_type.emplace_back(device); }
+			}
+			return devices_with_type;
+		} else
+			return devices;
+	}
+
+	template <cl::sycl::info::platform Param>
+	std::string get_info() const {
+		return name;
+	}
+
+	void set_info(std::string name) { this->name = name; }
+
+	bool operator!=(const mock_platform& other) const { return other.id == id; }
+
+	size_t get_id() { return id; }
+
+  private:
+	std::vector<mock_device> devices;
+	size_t id;
+	std::string name = "foo";
+};
+
+
+namespace celerity::detail {
+struct config_testspy {
+	static void set_mock_device_cfg(config& cfg, const device_config& d_cfg) { cfg.device_cfg = d_cfg; }
+	static void set_mock_host_cfg(config& cfg, const host_config& h_cfg) { cfg.host_cfg = h_cfg; }
+};
+} // namespace celerity::detail
+
+TEST_CASE_METHOD(celerity::test_utils::mpi_fixture, "pick_device prefers user specified device pointer", "[device-selection]") {
+	celerity::detail::config cfg(nullptr, nullptr);
+
+	mock_platform tp(68);
+	mock_device td(42, tp);
+	tp.set_devices({td});
+
+	auto device = pick_device(cfg, td, std::vector<mock_platform>{tp});
+	CHECK(device == td);
+}
+
+TEST_CASE_METHOD(celerity::test_utils::mpi_fixture,
+    "pick_device automatically selects a gpu device if available and otherwise falls back to the first device available", "[device-selection][gen]") {
```

Comment:
What is `[gen]`?

Reply:
Since the test used the "generator" for random combinations. Do you think it's superfluous?

Reply:
Yes, I'd say that's an implementation detail of the test case.